Source/SpatialGDK/Private/Utils/ComponentFactory.cpp

```diff
@@ -0,0 +1,288 @@
+		}
+		else
+		{
+			AddProperty(Object, Id, EnumProperty->GetUnderlyingProperty(), Data, UnresolvedObjects, ClearedIds);
+		}
+	}
+	else
+	{
+		checkNoEntry();
+		Schema_AddString(Object, Id, TEXT("Unknown Field"));
+	}
+}
+
+TArray<Worker_ComponentData> ComponentFactory::CreateComponentDatas(UObject* Object, const FPropertyChangeState& PropertyChangeState)
+{
+	TArray<Worker_ComponentData> ComponentData;
+
+	FClassInfo* Info = TypebindingManager->FindClassInfoByClass(Object->GetClass());
+	check(Info);
+
+	ComponentData.Add(CreateComponentData(Info->SingleClientComponent, PropertyChangeState, GROUP_SingleClient));
+	ComponentData.Add(CreateComponentData(Info->MultiClientComponent, PropertyChangeState, GROUP_MultiClient));
+
+	return ComponentData;
+}
+
+Worker_ComponentData ComponentFactory::CreateComponentData(Worker_ComponentId ComponentId, const struct FPropertyChangeState& Changes, EReplicatedPropertyGroup PropertyGroup)
+{
+	Worker_ComponentData ComponentData = {};
+	ComponentData.component_id = ComponentId;
+	ComponentData.schema_type = Schema_CreateComponentData(ComponentId);
+	Schema_Object* ComponentObject = Schema_GetComponentDataFields(ComponentData.schema_type);
+
+	FillSchemaObject(ComponentObject, Changes, PropertyGroup, true);
+
+	return ComponentData;
+}
+
+TArray<Worker_ComponentUpdate> ComponentFactory::CreateComponentUpdates(UObject* Object, const FPropertyChangeState& PropertyChangeState)
+{
+	TArray<Worker_ComponentUpdate> ComponentUpdates;
+
+	FClassInfo* Info = TypebindingManager->FindClassInfoByClass(Object->GetClass());
+	check(Info);
+
+	bool bWroteSomething = false;
+	Worker_ComponentUpdate SingleClientUpdate = CreateComponentUpdate(Info->SingleClientComponent, PropertyChangeState, GROUP_SingleClient, bWroteSomething);
+	if (bWroteSomething)
+	{
+		ComponentUpdates.Add(SingleClientUpdate);
+	}
+	bWroteSomething = false;
+	Worker_ComponentUpdate MultiClientUpdate = CreateComponentUpdate(Info->MultiClientComponent, PropertyChangeState, GROUP_MultiClient, bWroteSomething);
+	if (bWroteSomething)
+	{
+		ComponentUpdates.Add(MultiClientUpdate);
+	}
+
+	return ComponentUpdates;
+}
+
+Worker_ComponentUpdate ComponentFactory::CreateComponentUpdate(Worker_ComponentId ComponentId, const struct FPropertyChangeState& Changes, EReplicatedPropertyGroup PropertyGroup, bool& bWroteSomething)
+{
+	Worker_ComponentUpdate ComponentUpdate = {};
+
+	ComponentUpdate.component_id = ComponentId;
+	ComponentUpdate.schema_type = Schema_CreateComponentUpdate(ComponentId);
+	Schema_Object* ComponentObject = Schema_GetComponentUpdateFields(ComponentUpdate.schema_type);
+
+	TArray<Schema_FieldId> ClearedIds;
+
+	bWroteSomething = FillSchemaObject(ComponentObject, Changes, PropertyGroup, false, &ClearedIds);
+
+	for (Schema_FieldId Id : ClearedIds)
+	{
+		Schema_AddComponentUpdateClearedField(ComponentUpdate.schema_type, Id);
+	}
+
+	if (!bWroteSomething)
+	{
+		Schema_DestroyComponentUpdate(ComponentUpdate.schema_type);
```

Comment:
Be careful here. You're returning a valid `ComponentUpdate.schema_type` pointer, but you have already destroyed it. Best to assign it to `nullptr`.

Reply:
We don't pass this ComponentUpdate to the C API though.

Reply:
I know, but this code could change in the future, so it's better to not let stale pointers linger about.